File: harness/adapters/typescript/adapter.ts
// Reference npamp_00 conformance adapter (TypeScript). A third-language "testee" proving
// the runner is language-agnostic. Reads length-prefixed JSON requests {op,in} on stdin,
// writes length-prefixed JSON responses {out|error|skipped} on stdout, performing the real
// npamp_00 primitive for each op by CALLING the impl/typescript reference implementation
// (src/npamp.ts) -- it does not reimplement the primitives.
//
// Windows note: reads/writes the binary stdio streams (process.stdin/stdout are binary in
// Node; no CRLF translation is applied to Buffers) and FLUSHES nothing extra is needed since
// Node flushes Buffer writes, but we wait for the drain/callback to guarantee ordering.
//
// Mapping notes (adapter -> reference):
//  - crc32c(octets)        -> crc32c(buf) from src/npamp.ts (number -> 4-byte big-endian hex)
//  - header.encode(fields) -> new Frame({...}).marshal()  (frame is header(36) + payload(0))
//  - header.decode(frame)  -> Frame.unmarshal(buf); crc/fields read back from the buffer
//  - aead.seal/open        -> sealAes256Gcm/openAes256Gcm(key, iv=nonce, seq=0n, aad, pt|sealed)
//                             with seq=0 the impl's deriveNonce(iv,0) == iv, so the raw nonce
//                             from the corpus is used verbatim on the REAL seal/open path.
//  - hkdf.expand           -> hkdfExpand(hash, prk, info, length) from src/npamp.ts
//  - tlv.decode            -> wire TLV parse (the reference impl ships no TLV decoder)
//  - profile.check         -> skipped (the reference impl ships no profile checker)
import { writeSync } from "node:fs";
import {
  crc32c,
  Frame,
  sealAes256Gcm,
  openAes256Gcm,
  hkdfExpand,
  HEADER_SIZE,
  MAGIC,
} from "../../../impl/typescript/src/npamp.ts";

const BREAK = process.argv.slice(2).includes("--break");

type Req = { op?: string; in?: Record<string, unknown> };
type Resp =
  | { out: Record<string, unknown> }
  | { error: string }
  | { skipped: string };

function s(o: Record<string, unknown>, k: string): string {
  const v = o[k];
  return typeof v === "string" ? v : "";
}
function num(o: Record<string, unknown>, k: string): number {
  const v = o[k];
  return typeof v === "number" ? v : Number(v ?? 0);
}
function hx(o: Record<string, unknown>, k: string): Buffer {
  return Buffer.from(s(o, k), "hex");
}
function u32hex(n: number): string {
  const b = Buffer.alloc(4);
  b.writeUInt32BE(n >>> 0, 0);
  return b.toString("hex");
}

function handle(req: Req): Resp {
  const op = req.op;
  const i = req.in ?? {};
  switch (op) {
    case "header.encode": {
      // Build the frame from its fields via the reference Frame + marshal().
      const f = new Frame({
        version: num(i, "ver"),
        flags: num(i, "flags"),
        ftype: num(i, "frameType"),
        channel: num(i, "channel"),
        seq: BigInt(num(i, "seq")),
        // payload is empty for header vectors; payloadLength encodes from payload.length,
        // and the golden vectors carry payloadLength 0.
      });
      return { out: { frame: f.marshal().toString("hex") } };
    }

    case "header.decode": {
      const b = hx(i, "frame");
      if (b.length !== HEADER_SIZE) return { error: "malformed header length" };
      let f: Frame;
      try {
        f = Frame.unmarshal(b);
      } catch (e) {
        return { error: (e as Error).message };
      }
      return {
        out: {
          magic: MAGIC.toString("latin1"),
          ver: f.version,
          flags: f.flags,
          frameType: f.ftype,
          channel: f.channel,
          seq: Number(f.seq),
          payloadLength: b.readUInt32BE(17),
          crc32c: b.subarray(21, 25).toString("hex"),
          reservedZero: true,
        },
      };
    }

    case "crc32c": {
      const b = hx(i, "octets");
      let c = u32hex(crc32c(b));
      if (BREAK) c = "deadbeef"; // deliberate corruption for the runner's mutation check
      return { out: { crc32c: c } };
    }

    case "tlv.decode": {
      const b = hx(i, "tlv");
      if (b.length < 4) return { error: "truncated tlv" };
      const typ = b.readUInt16BE(0);
      const length = b.readUInt16BE(2);
      if (typ & 0x8000)
        return { error: "unknown forward-incompatible TLV (high bit set)" };
      if (length !== b.length - 4) return { error: "tlv length mismatch" };
      return {
        out: { type: typ, length, value: b.subarray(4).toString("hex") },
      };
    }

    case "aead.seal": {
      if (s(i, "suite") !== "AES-256-GCM")
        return { skipped: "suite not implemented: " + s(i, "suite") };
      try {
        // seq=0n -> deriveNonce(iv,0)==iv, so the corpus nonce is used as-is.
        const sealed = sealAes256Gcm(
          hx(i, "key"),
          hx(i, "nonce"),
          0n,
          hx(i, "aad"),
          hx(i, "pt"),
        );
        return { out: { sealed: sealed.toString("hex") } };
      } catch (e) {
        return { error: (e as Error).message };
      }
    }

    case "aead.open": {
      if (s(i, "suite") !== "AES-256-GCM")
        return { skipped: "suite not implemented: " + s(i, "suite") };
      try {
        const pt = openAes256Gcm(
          hx(i, "key"),
          hx(i, "nonce"),
          0n,
          hx(i, "aad"),
          hx(i, "sealed"),
        );
        return { out: { pt: pt.toString("hex") } };
      } catch {
        return { error: "authentication failed" };
      }
    }

    case "hkdf.expand": {
      const hash = s(i, "hash");
      if (hash !== "sha256" && hash !== "sha384")
        return { skipped: "hash not implemented: " + hash };
      const okm = hkdfExpand(hash, hx(i, "prk"), hx(i, "info"), num(i, "length"));
      return { out: { okm: okm.toString("hex") } };
    }

    case "profile.check":
      // The OPEN reference implementation ships no profile-acceptance checker.
      return { skipped: "profile.check not implemented in reference impl" };

    default:
      return { skipped: "op not implemented: " + op };
  }
}

// ---- length-prefixed framing over binary stdin/stdout ----
// 4-byte little-endian length N, then N bytes of JSON. Repeat until stdin closes.

function writeResp(resp: Resp): void {
  const ob = Buffer.from(JSON.stringify(resp), "utf8");
  const ol = Buffer.alloc(4);
  ol.writeUInt32LE(ob.length, 0);
  // Synchronous write to fd 1 guarantees the bytes are flushed and ordered on Windows,
  // so the length prefix and JSON body are never split or reordered across responses.
  writeSync(1, ol);
  writeSync(1, ob);
}

function main(): void {
  let buffered = Buffer.alloc(0);

  function drain(): void {
    for (;;) {
      if (buffered.length < 4) return;
      const n = buffered.readUInt32LE(0);
      if (buffered.length < 4 + n) return;
      const body = buffered.subarray(4, 4 + n);
      buffered = buffered.subarray(4 + n);
      let resp: Resp;
      try {
        resp = handle(JSON.parse(body.toString("utf8")) as Req);
      } catch (e) {
        resp = { error: "adapter exception: " + (e as Error).message };
      }
      writeResp(resp);
    }
  }

  process.stdin.on("data", (c: Buffer) => {
    buffered = Buffer.concat([buffered, c]);
    drain();
  });
  process.stdin.on("end", () => process.exit(0));
  process.stdin.on("close", () => process.exit(0));
}

main();

File: harness/adapters/python/adapter.py
"""Reference npamp_00 conformance adapter (Python). A second-language "testee" proving
the runner is language-agnostic. Reads length-prefixed JSON requests {op,in} on stdin,
writes length-prefixed JSON responses {out|error|skipped} on stdout, performing the real
npamp_00 primitive for each op.

Windows note: uses the binary stdio buffers (text mode would corrupt the byte framing via
CRLF translation) and flushes after every response. Requires the `cryptography` package
for AES-256-GCM. Flag --break corrupts crc32c for the runner's mutation check.
"""
import hashlib
import hmac
import json
import struct
import sys

from cryptography.hazmat.primitives.ciphers.aead import AESGCM

BREAK = "--break" in sys.argv[1:]


def _crc32c_table():
    poly = 0x82F63B78  # reflected Castagnoli
    tbl = []
    for i in range(256):
        crc = i
        for _ in range(8):
            crc = (crc >> 1) ^ (poly if (crc & 1) else 0)
        tbl.append(crc & 0xFFFFFFFF)
    return tbl


_TBL = _crc32c_table()


def crc32c(data: bytes) -> int:
    crc = 0xFFFFFFFF
    for b in data:
        crc = (crc >> 8) ^ _TBL[(crc ^ b) & 0xFF]
    return crc ^ 0xFFFFFFFF


def crc32c_hex(data: bytes) -> str:
    return crc32c(data).to_bytes(4, "big").hex()


def hkdf_expand(hashname, prk, info, length):
    h = hashlib.sha256 if hashname == "sha256" else hashlib.sha384
    out, t, i = b"", b"", 1
    while len(out) < length:
        t = hmac.new(prk, t + info + bytes([i]), h).digest()
        out += t
        i += 1
    return out[:length]


def hx(d, k):
    return bytes.fromhex(d.get(k, "") or "")


def handle(req):
    op = req.get("op")
    i = req.get("in", {})
    if op == "header.encode":
        h = bytearray(36)
        h[0:4] = b"NPAM"
        h[4] = ((int(i["ver"]) & 0x0F) << 4) | (int(i["flags"]) & 0x0F)
        h[5:7] = int(i["frameType"]).to_bytes(2, "big")
        h[7:9] = int(i["channel"]).to_bytes(2, "big")
        h[9:17] = int(i["seq"]).to_bytes(8, "big")
        h[17:21] = int(i["payloadLength"]).to_bytes(4, "big")
        h[21:25] = crc32c(bytes(h[0:21])).to_bytes(4, "big")
        return {"out": {"frame": bytes(h).hex()}}
    if op == "header.decode":
        b = hx(i, "frame")
        if len(b) != 36:
            return {"error": "malformed header length"}
        if b[0:4] != b"NPAM":
            return {"error": "bad magic"}
        if b[21:25] != crc32c(b[0:21]).to_bytes(4, "big"):
            return {"error": "crc32c mismatch"}
        if any(x != 0 for x in b[25:36]):
            return {"error": "reserved octet non-zero"}
        return {"out": {
            "magic": "NPAM", "ver": b[4] >> 4, "flags": b[4] & 0x0F,
            "frameType": int.from_bytes(b[5:7], "big"),
            "channel": int.from_bytes(b[7:9], "big"),
            "seq": int.from_bytes(b[9:17], "big"),
            "payloadLength": int.from_bytes(b[17:21], "big"),
            "crc32c": b[21:25].hex(), "reservedZero": True}}
    if op == "crc32c":
        c = "deadbeef" if BREAK else crc32c_hex(hx(i, "octets"))
        return {"out": {"crc32c": c}}
    if op == "tlv.decode":
        b = hx(i, "tlv")
        if len(b) < 4:
            return {"error": "truncated tlv"}
        typ = int.from_bytes(b[0:2], "big")
        length = int.from_bytes(b[2:4], "big")
        if typ & 0x8000:
            return {"error": "unknown forward-incompatible TLV (high bit set)"}
        if length != len(b) - 4:
            return {"error": "tlv length mismatch"}
        return {"out": {"type": typ, "length": length, "value": b[4:].hex()}}
    if op == "aead.seal":
        if i.get("suite") != "AES-256-GCM":
            return {"skipped": "suite not implemented: %s" % i.get("suite")}
        try:
            sealed = AESGCM(hx(i, "key")).encrypt(hx(i, "nonce"), hx(i, "pt"), hx(i, "aad"))
        except Exception as e:
            return {"error": str(e)}
        return {"out": {"sealed": sealed.hex()}}
    if op == "aead.open":
        if i.get("suite") != "AES-256-GCM":
            return {"skipped": "suite not implemented: %s" % i.get("suite")}
        try:
            pt = AESGCM(hx(i, "key")).decrypt(hx(i, "nonce"), hx(i, "sealed"), hx(i, "aad"))
        except Exception:
            return {"error": "authentication failed"}
        return {"out": {"pt": pt.hex()}}
    if op == "hkdf.expand":
        hn = i.get("hash")
        if hn not in ("sha256", "sha384"):
            return {"skipped": "hash not implemented: %s" % hn}
        return {"out": {"okm": hkdf_expand(hn, hx(i, "prk"), hx(i, "info"), int(i["length"])).hex()}}
    if op == "profile.check":
        profile, kem = i.get("profile"), i.get("kem")
        if profile == "Sovereign" and kem == "X25519MLKEM768":
            return {"error": "Sovereign MUST NOT accept X25519MLKEM768"}
        if profile == "High" and kem == "X25519MLKEM768":
            return {"error": "High minimum KEM is X25519MLKEM1024"}
        return {"out": {"accepted": True}}
    return {"skipped": "op not implemented: %s" % op}


def main():
    rd, wr = sys.stdin.buffer, sys.stdout.buffer
    while True:
        lp = rd.read(4)
        if len(lp) < 4:
            return
        n = struct.unpack("<I", lp)[0]
        body = rd.read(n)
        try:
            resp = handle(json.loads(body))
        except Exception as e:  # noqa: BLE001
            resp = {"error": "adapter exception: %s" % e}
        ob = json.dumps(resp).encode()
        wr.write(struct.pack("<I", len(ob)))
        wr.write(ob)
        wr.flush()


if __name__ == "__main__":
    main()
